File: ui/src/lib/format.ts
// Small display helpers — no protocol knowledge here.

export function shortId(id: string): string {
  const raw = id.includes(':') ? id.slice(id.indexOf(':') + 1) : id;
  if (raw.length <= 10) return raw;
  return `${raw.slice(0, 4)}…${raw.slice(-4)}`;
}

export function prettyLabel(label: string): string {
  const s = label.replace(/[_-]+/g, ' ').trim();
  return s ? s.charAt(0).toUpperCase() + s.slice(1) : label;
}

/** Tone for an agent-status label. Shared so the timeline chip and the Agents
 *  panel dot agree.
 *
 *  Green is earned, never the fallback: only labels carrying a known
 *  healthy/active token render the accent. Labels are free-form wire data —
 *  a label this English-token contract can't read (including any non-English
 *  label, e.g. `échec`) must render neutral, because guessing green for it
 *  would be exactly the reassuring-but-untrue state the honesty rules forbid.
 *  The token contract is documented in docs/agent-guide.md. */
const GREEN_TOKENS =
  /\b(done|working|online|ready|pass|passed|success|successful|ok|complete|completed|connected|healthy|active|running|verified|live)\b/;
// Word-boundary tokens, not substrings: `review` must not match inside
// `preview` (preview_ready is a success label, not a waiting one).
const BLUE_TOKENS = /\b(await(ing)?|review(ing|ed)?|pend(ing)?)\b/;

export function labelTone(label: string): 'red' | 'blue' | 'green' | 'neutral' {
  const l = label.toLowerCase().replace(/[_-]+/g, ' ');
  // Substrings on purpose for red: a false alarm is the honest direction to
  // err in — never the reverse.
  if (l.includes('fail') || l.includes('error') || l.includes('block')) return 'red';
  if (BLUE_TOKENS.test(l)) return 'blue';
  if (GREEN_TOKENS.test(l)) return 'green';
  return 'neutral';
}

// Must stay in sync with the corresponding tokens in styles.css.
const ACCENT = '#2fd6a4'; // --accent
const BLUE = '#6aa8f7'; // --blue
const RED = '#f26d6d'; // --red
const TEXT_DIM = '#8aa39d'; // --text-dim
const VIOLET = '#a78bfa'; // no CSS token — shared here by AVATAR_PALETTE and fileTint

const AVATAR_PALETTE = [ACCENT, BLUE, VIOLET, '#fb923c', '#f472b6', '#22d3ee'];

export function colorForId(id: string): string {
  let h = 0;
  for (let i = 0; i < id.length; i++) h = (Math.imul(h, 31) + id.charCodeAt(i)) >>> 0;
  return AVATAR_PALETTE[h % AVATAR_PALETTE.length];
}

export function initials(name: string): string {
  const parts = name.split(/\s+/).filter(Boolean);
  if (parts.length === 0) return '?';
  if (parts.length === 1) return parts[0].slice(0, 2).toUpperCase();
  return (parts[0].charAt(0) + parts[1].charAt(0)).toUpperCase();
}

export function extOf(name: string): string {
  const i = name.lastIndexOf('.');
  return i >= 0 ? name.slice(i + 1).toLowerCase() : '';
}

export function fileTint(name: string): string {
  switch (extOf(name)) {
    case 'pdf':
      return RED;
    case 'md':
    case 'txt':
    case 'doc':
    case 'docx':
      return BLUE;
    case 'json':
    case 'js':
    case 'ts':
      return ACCENT;
    case 'png':
    case 'jpg':
    case 'jpeg':
    case 'gif':
    case 'svg':
    case 'webp':
      return VIOLET;
    default:
      return TEXT_DIM;
  }
}

File: dart/jeliya_protocol/lib/src/conventions/format.dart
/// Display-classification helpers ported verbatim from ui/src/lib/format.ts —
/// only the pieces the protocol conventions themselves depend on ([labelTone]
/// per docs/PROTOCOL.md "Agent-status label tone", and [shortId] which the
/// diagnostics report uses to truncate identifiers). Purely visual formatting
/// (bytes, times, avatars) stays in the app layer.
library;

/// Tone classes for an agent-status label. Derived, never wire data.
enum LabelTone { red, blue, green, neutral }

// Green is earned, never the fallback: only labels carrying a known
// healthy/active token render the accent (honesty rule 4). The token contract
// is documented in docs/agent-guide.md.
final RegExp _greenTokens = RegExp(
    r'\b(done|working|online|ready|pass|passed|success|successful|ok|complete|completed'
    r'|connected|healthy|active|running|verified|live)\b');
// Word-boundary tokens, not substrings: `review` must not match inside
// `preview` (preview_ready is a success label, not a waiting one).
final RegExp _blueTokens = RegExp(r'\b(await(ing)?|review(ing|ed)?|pend(ing)?)\b');
final RegExp _separators = RegExp('[_-]+');

/// Tone for an agent-status label (format.ts `labelTone`) — shared so the
/// timeline chip and the Agents panel dot agree, and mirrored exactly across
/// clients (docs/PROTOCOL.md "Agent-status label tone").
///
/// Applied to the label lowercased with `_`/`-` collapsed to spaces, in this
/// precedence: red substrings (`fail`/`error`/`block` — substrings on purpose,
/// a false alarm is the honest direction to err in), then blue word-boundary
/// tokens, then green word-boundary tokens, else neutral. Labels are free-form
/// wire data — a label this English-token contract can't read (including any
/// non-English label, e.g. `échec`) must render neutral, because guessing
/// green for it would be exactly the reassuring-but-untrue state the honesty
/// rules forbid.
LabelTone labelTone(String label) {
  final l = label.toLowerCase().replaceAll(_separators, ' ');
  if (l.contains('fail') || l.contains('error') || l.contains('block')) return LabelTone.red;
  if (_blueTokens.hasMatch(l)) return LabelTone.blue;
  if (_greenTokens.hasMatch(l)) return LabelTone.green;
  return LabelTone.neutral;
}

/// Truncate an identifier for display (format.ts `shortId`): drop any
/// `<scheme>:` prefix, then keep 4+4 chars around an ellipsis once the raw id
/// exceeds 10 chars.
String shortId(String id) {
  final raw = id.contains(':') ? id.substring(id.indexOf(':') + 1) : id;
  if (raw.length <= 10) return raw;
  return '${raw.substring(0, 4)}…${raw.substring(raw.length - 4)}';
}
